### PR TITLE
feat: metamask support

### PR DESCRIPTION
**Pull request by mvpratt**
## Description

Please describe your changes

## Notice

Before submitting a pull request, please make sure you have answered the following:

- [ ] Have you followed the guidelines in our [Contributing]('https://github.com/shapeshift/web/CONTRIBUTING.md) guide?
- [ ] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/shapeshift/web/pulls) for the same update/change?
- [ ] Do all new and existing tests pass? Does the linter pass?

## Pull Request Type

- [ ] :bug: Bug fix (Non-breaking Change: Fixes an issue)
- [ ] :hammer_and_wrench: Chore (Non-breaking Change: Doc updates, pkg upgrades, typos, etc..)
- [ ] :nail_care: New Feature (Breaking/Non-breaking Change)

## Issue (if applicable)

If applicable, please link to the github issue and put `closes #XXXX` in your comment to auto-close the issue that your PR fixes.

## Testing

Please outline all testing steps

1. Pull branch locally and run `yarn` to install new deps
2. etc...

## Screenshots (if applicable)
Screen Shot 2021-10-28 at 12 58 15 PM

Screen Shot 2021-10-28 at 12 57 49 PM

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -31,6 +31,7 @@
     "bip39": "^3.0.4",
     "crypto-js": "^4.1.1",
     "dayjs": "^1.10.6",
+    "@metamask/detect-provider": "^1.2.0",
```
Comment (cjthompson) [inline package.json:34]: does this not belong in `hdwallet-metamask`?

**Review by cjthompson — APPROVED** [review]

File: src/components/Modals/Send/hooks/useFormSend/useFormSend.tsx [hunk 2]
```diff
@@ -46,9 +46,26 @@ export const useFormSend = () => {
...
+          await adapter.broadcastTransaction(signedTx)
+        } else if (wallet.supportsBroadcast()) {
+          /**
+           * signAndBroadcastTransaction is an optional method on the HDWallet interface.
+           * Check and see if it exists; if so, call and make sure a txhash is returned
+           */
+          if (!adapter.signAndBroadcastTransaction) {
+            throw new Error('signAndBroadcastTransaction undefined for wallet')
+          }
+          broadcastTXID = await adapter.signAndBroadcastTransaction?.({ txToSign, wallet })
```
Comment (0xdef1cafe) [inline src/components/Modals/Send/hooks/useFormSend/useFormSend.tsx:62]: you could define as const here as it looks like this was done to get things working as we never reassign

File: src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.test.ts [hunk 3]
```diff
@@ -6,10 +6,10 @@ import { selectAndSortAssets } from './selectAndSortAssets'
...
   ...mockStore,
   assets: {
-    [rune.tokenId]: rune,
-    [aapl.tokenId]: aapl,
-    [ethereum.chain]: ethereum,
-    [zero.tokenId]: zero
+    [rune.tokenId as string]: rune,
+    [aapl.tokenId as string]: aapl,
+    [ethereum.chain as string]: ethereum,
+    [zero.tokenId as string]: zero
```
Comment (0xdef1cafe) [inline src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.test.ts:12]: better fix is to give the assets in `jest/mock/assets` the `Asset` type and we don't have to do this casting everywhere and the tests will be more robust

File: src/components/Modals/Send/hooks/useFormSend/useFormSend.tsx (hunk 2 above)
Comment (pastaghost) [reply src/components/Modals/Send/hooks/useFormSend/useFormSend.tsx:62]: We're probably going to need this return value. I intended to pass this to the toast a few lines below, but that isn't yet implemented.

File: src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.test.ts (hunk 3 above)
Comment (pastaghost) [reply src/components/AssetSearch/selectors/selectAndSortAssets/selectAndSortAssets.test.ts:12]: I did give the assets in `jest/mock/assets` the `Asset` type. Without the casting, tsc complains here about the `.tokenId` fields being of type `string | undefined` when a key of type `string` is required.

**Review by 0xdef1cafe — APPROVED** [review]

**Pull request merged**